In [1]:
import pandas as pd

In [15]:
df=pd.read_csv(r"https://raw.githubusercontent.com/sunnysavita10/NLP-With-Machine-Learning/main/SpamClassifier-with-ML/sms_spam_data/SMSSpamCollection.csv",sep="\t",header=None,names=["label","messages"])

In [16]:
df

label                                           messages
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [17]:
df.dtypes

label       object
messages    object
dtype: object

it is neither cateogarical nor int its ustructured text data

In [26]:
len(df)

5572

In [57]:
df['messages'][0]

'Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'

## WE will used NAIVE BAYES for it

In [18]:
import nltk

In [19]:
import re #it is regular expression used to find pattern in text data

In [20]:
nltk.download("stopwords") #it contain few words which are not important wehich we will remove from ourr text

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\harsh\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [21]:
from nltk.corpus import stopwords

In [29]:
stopwords.words('english')

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [22]:
from nltk.stem.porter import PorterStemmer #to get the root words of simliar words

In [24]:
ps=PorterStemmer()

We will do some preproceesiong
- Removing , . capital letters to lowe letters and etc
- then we will encode our lettters

In [58]:
corpus = []
for i in range(len(df)):
    # Step 1: Remove non-alphabetic characters
    review = re.sub('[^a-zA-Z]', ' ', df["messages"][i])
    print(f"Step 1 - Non-alphabetic characters removed: {review}")
    
    # Step 2: Convert to lowercase
    review = review.lower()
    print(f"Step 2 - Converted to lowercase: {review}")
    
    # Step 3: Split into words
    review = review.split()
    print(f"Step 3 - Split into words: {review}")
    
    # Step 4: Remove stopwords and stem the words
    review = [ps.stem(word) for word in review if word not in stopwords.words('english')]
    print(f"Step 4 - Stopwords removed and words stemmed: {review}")
    
    # Step 5: Join the words back into a single string
    review = ' '.join(review)
    print(f"Step 5 - Words joined: {review}")
    
    # Step 6: Add to the corpus
    corpus.append(review)

# Output the processed corpus
print("Final corpus:")
print(corpus)


Step 1 - Non-alphabetic characters removed: Go until jurong point  crazy   Available only in bugis n great world la e buffet    Cine there got amore wat   
Step 2 - Converted to lowercase: go until jurong point  crazy   available only in bugis n great world la e buffet    cine there got amore wat   
Step 3 - Split into words: ['go', 'until', 'jurong', 'point', 'crazy', 'available', 'only', 'in', 'bugis', 'n', 'great', 'world', 'la', 'e', 'buffet', 'cine', 'there', 'got', 'amore', 'wat']
Step 4 - Stopwords removed and words stemmed: ['go', 'jurong', 'point', 'crazi', 'avail', 'bugi', 'n', 'great', 'world', 'la', 'e', 'buffet', 'cine', 'got', 'amor', 'wat']
Step 5 - Words joined: go jurong point crazi avail bugi n great world la e buffet cine got amor wat
Step 1 - Non-alphabetic characters removed: Ok lar    Joking wif u oni   
Step 2 - Converted to lowercase: ok lar    joking wif u oni   
Step 3 - Split into words: ['ok', 'lar', 'joking', 'wif', 'u', 'oni']
Step 4 - Stopwords removed an

In [59]:
corpus

['go jurong point crazi avail bugi n great world la e buffet cine got amor wat',
 'ok lar joke wif u oni',
 'free entri wkli comp win fa cup final tkt st may text fa receiv entri question std txt rate c appli',
 'u dun say earli hor u c alreadi say',
 'nah think goe usf live around though',
 'freemsg hey darl week word back like fun still tb ok xxx std chg send rcv',
 'even brother like speak treat like aid patent',
 'per request mell mell oru minnaminungint nurungu vettam set callertun caller press copi friend callertun',
 'winner valu network custom select receivea prize reward claim call claim code kl valid hour',
 'mobil month u r entitl updat latest colour mobil camera free call mobil updat co free',
 'gonna home soon want talk stuff anymor tonight k cri enough today',
 'six chanc win cash pound txt csh send cost p day day tsandc appli repli hl info',
 'urgent week free membership prize jackpot txt word claim c www dbuk net lccltd pobox ldnw rw',
 'search right word thank breather

In [60]:
from sklearn.feature_extraction.text import CountVectorizer

In [61]:
cv = CountVectorizer()

In [62]:
X=cv.fit_transform(corpus).toarray()

In [63]:
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [64]:
len(X)

5572

In [65]:
X.ndim

2

In [66]:
X.shape

(5572, 6296)

## Too many columns and lines
- So if we will train on this data our model will be overfitted as 6296 columns means 6296 unique words and
- With respect to each row it checking whether all 6296 words are there or not if words prestnt 
- then 1 for present in that place of uniquies words otherwise 0
- so we will take columns which has maximium 1 so that we can reduce columns and retain much info.

In [67]:
cv1=CountVectorizer(max_features=2500) #lets take 2500 randomly

In [68]:
X1=cv1.fit_transform(corpus).toarray()

In [69]:
X1

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [70]:
X1.shape

(5572, 2500)

### now for target column

In [71]:
df["label"]

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: label, Length: 5572, dtype: object

In [77]:
y=pd.get_dummies(df['label'],dtype=int,drop_first=True)

In [78]:
y

spam
0        0
1        0
2        1
3        0
4        0
...    ...
5567     1
5568     0
5569     0
5570     0
5571     0

[5572 rows x 1 columns]

#### now we will create model

In [79]:
from sklearn.model_selection import train_test_split

In [82]:
X_train,X_test,y_train,y_test=train_test_split(X1,y,random_state=42,test_size=25)

In [83]:
from sklearn.naive_bayes import GaussianNB

In [84]:
model=GaussianNB()

In [85]:
model.fit(X_train,y_train)

C:\Users\harsh\anaconda3\Lib\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GaussianNB()

In [87]:
y_pred=model.predict(X_test)

In [88]:
from sklearn.metrics import accuracy_score

In [89]:
accuracy_score(y_test,y_pred)

0.88

In [90]:
from sklearn.naive_bayes import MultinomialNB

In [91]:
model2=MultinomialNB()

In [92]:
model2.fit(X_train,y_train)

C:\Users\harsh\anaconda3\Lib\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MultinomialNB()

In [94]:
y1_pred=model2.predict(X_test)

In [95]:
accuracy_score(y_test,y1_pred)

0.96